src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/BulkCopyTestUtil.java

```diff
@@ -86,6 +174,137 @@ static void performBulkCopy(BulkCopyTestWrapper wrapper, DBTable sourceTable, bo
+            }
+        }
+        finally {
+            stmt.dropTable(destinationTable);
+            con.close();
+        }
+    }
+    
+    /**
+     * perform bulk copy using source and destination tables
+     * 
+     * @param wrapper
+     * @param sourceTable
+     * @param destTable
+     * @param validateResult
+     * @param fail
+     * @param dropDest
+     */
+    static void performBulkCopy(BulkCopyTestWrapper wrapper, DBTable sourceTable, DBTable destinationTable, boolean validateResult, boolean fail, boolean dropDest) {
+        DBConnection con = null;
+        DBStatement stmt = null;
+        try {
+            con = new DBConnection(wrapper.getConnectionString());
+            stmt = con.createStatement();
+
+            DBResultSet srcResultSet = stmt.executeQuery("SELECT * FROM " + sourceTable.getEscapedTableName() + ";");
+            SQLServerBulkCopy bulkCopy;
+            if (wrapper.isUsingConnection()) {
+                bulkCopy = new SQLServerBulkCopy((Connection) con.product());
+            }
+            else {
+                bulkCopy = new SQLServerBulkCopy(wrapper.getConnectionString());
+            }
+            if (wrapper.isUsingBulkCopyOptions()) {
+                bulkCopy.setBulkCopyOptions(wrapper.getBulkOptions());
+            }
+            bulkCopy.setDestinationTableName(destinationTable.getEscapedTableName());
+            if (wrapper.isUsingColumnMapping()){
+                for(int i = 0; i < wrapper.cm.size(); i++){
+                     ColumnMap currentMap = wrapper.cm.get(i);
+                     if(currentMap.sourceIsInt && currentMap.destIsInt){
+                         bulkCopy.addColumnMapping(currentMap.srcInt, currentMap.destInt);
+                     }
+                     else if(currentMap.sourceIsInt && (!currentMap.destIsInt)){
+                         bulkCopy.addColumnMapping(currentMap.srcInt, currentMap.destString);
+                     }
+                     else if((!currentMap.sourceIsInt) && currentMap.destIsInt){
+                         bulkCopy.addColumnMapping(currentMap.srcString, currentMap.destInt);
+                     }
+                     else if((!currentMap.sourceIsInt) && (!currentMap.destIsInt)){
+                         bulkCopy.addColumnMapping(currentMap.srcString, currentMap.destString);
+                     }
+                }
+            }
+            bulkCopy.writeToServer((ResultSet) srcResultSet.product());
+            if(fail)
+                fail("bulkCopy.writeToServer did not fail when it should have");
+            bulkCopy.close();
+            if (validateResult) {
+                validateValues(con, sourceTable, destinationTable);
+            }
+        }
+        catch (SQLException ex) {
+            if(!fail){
+                fail(ex.getMessage());
+            }
+        }
+        finally {
+            if(dropDest){
+                stmt.dropTable(destinationTable);
+                con.close();
```

Comment:
The if(dropDest) condition is because some destination tables require further validation after performBulkCopy method.

i've moved con.close() to outside the if statement